src/seraphis_mocks/scan_context_async_mock.cpp

```diff
@@ -0,0 +1,797 @@
+    do
+    {
+        ++try_count;
+        try
+        {
+            MDEBUG("Pulling blocks at req start height: " << req.start_height << " (try_count=" << try_count << ")");
+            r = rpc_get_blocks(req, res_out);
+            const std::string status = cryptonote::get_rpc_status(trusted_daemon, res_out.status);
+            THROW_ON_RPC_RESPONSE_ERROR(r, {}, res_out, "getblocks.bin", tools::error::get_blocks_error, status);
+            validate_get_blocks_res(chunk_request, res_out);
+        }
+        catch (tools::error::deprecated_rpc_access&)
+        {
+            // No need to retry
+            std::rethrow_exception(std::current_exception());
+        }
+        catch (...)
+        {
+            r = false;
+            if (try_count >= max_get_blocks_attempts)
+                std::rethrow_exception(std::current_exception());
+        }
+    } while (!r && try_count < max_get_blocks_attempts);
+
+    THROW_WALLET_EXCEPTION_IF(!r, tools::error::wallet_internal_error, "failed to get blocks");
+
+    MDEBUG("Pulled blocks: requested start height " << req.start_height << ", count " << res_out.blocks.size()
+        << ", node height " << res_out.current_height << ", top hash " << res_out.top_block_hash
+        << ", pool info " << static_cast<unsigned int>(res_out.pool_info_extent));
+}
+//-------------------------------------------------------------------------------------------------------------------
+//-------------------------------------------------------------------------------------------------------------------
+static void parse_rpc_get_blocks(const ChunkRequest &chunk_request,
+    const cryptonote::COMMAND_RPC_GET_BLOCKS_FAST::response &res,
+    sp::scanning::ChunkContext &chunk_context_out,
+    sp::LegacyUnscannedChunk &unscanned_chunk_out)
+{
+    validate_get_blocks_res(chunk_request, res);
+
+    // Older daemons can return more blocks than requested because they did not support a max_block_count req param.
+    // The scanner expects requested_chunk_size blocks however, so we only care about the blocks up until that point.
+    // Note the scanner can also return *fewer* blocks than requested if at chain tip or the chunk exceeded max size.
+    const std::uint64_t num_blocks = std::min((std::uint64_t)res.blocks.size(), chunk_request.requested_chunk_size);
+
+    chunk_context_out.block_ids.clear();
+
+    unscanned_chunk_out.clear();
+    unscanned_chunk_out.resize(num_blocks);
+
+    if (num_blocks == 0)
+    {
+        // must have requested the tip of the chain
+        chunk_context_out.prefix_block_id = rct::hash2rct(res.top_block_hash);
+        chunk_context_out.start_index = res.current_height;
+        return;
+    }
+
+    // parse blocks and txs
+    for (std::size_t block_idx = 0; block_idx < num_blocks; ++block_idx)
+    {
+        auto &unscanned_block = unscanned_chunk_out[block_idx];
+        unscanned_block.unscanned_txs.resize(1 + res.blocks[block_idx].txs.size());
```

Comment:
```suggestion
        const auto &res_block = res.blocks[block_idx];
        auto &unscanned_block = unscanned_chunk_out[block_idx];
        unscanned_block.unscanned_txs.resize(1 + res_block.txs.size());
```

Either this or write a separate function for the loop body (imo it's big enough for a new function).